### ISSUE TITLE
Head field references?

### ISSUE THREAD

**Issue by jsgf** [joshua-maros/ouroboros#13]
I have a structure of the form
```
#[self_referencing]
struct MyStruct {
    foo: Arc<Foo>,
    #[borrows(foo)]
    bar: Thing<'this>,
}
```
I want to be able to create new clones of `Arc<Foo>`, but it seems I can only get its contents with `borrow`/`with_foo_contents()`. Is there a soundness problem with having `fn borrow_foo(&self) -> &Arc<Foo>` so I can clone the Arc?

**Comment by joshua-maros** [comment]
I was uncertain if there was a soundness issue, so I opted to only make the contents available to be safe. A workaround would be to make foo type `Box<Arc<Foo>>` which you should still be able to construct `Thing<'this>` from by dereferencing it, although that would be a less elegant solution than being able to borrow `foo` directly. 

I'm pretty sure that in the case of `Box`, `Rc`, and `Arc`, there is no soundness issue with being able to borrow the entire field. However, I do not know if that holds true for all types that implement `StableDeref`.

**Comment by jsgf** [comment]
My only concern was `Arc::make_mut`/`Arc::get_mut`, but since that takes `&mut self` it can't be a problem here. I think that's going to be true for any type which doesn't have interior mutability, and I don't think a type with interior mutability can safely implement `StableDeref`.

**Comment by joshua-maros** [comment]
That is a sound argument. I have changed the `borrow_*_contents` and `with_*_contents` functions to just be `borrow_*` and `with_*`, and published the changes as version `0.7.0`. I also made `chain_hack` work with `Arc` and `Rc` in addition to `Box` while doing so, realizing it would be trivial.
